File: src/WallpaperReddit/Services/Library.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace WallpaperReddit.Services
{
    /// <summary>
    /// The wallpaper history plus the blacklist, persisted as JSON. Small data set
    /// (hundreds of rows at most), so a flat file keeps the app dependency-free.
    /// All public methods are thread-safe.
    /// </summary>
    public class Library
    {
        private static readonly JsonSerializerOptions Options = new()
        {
            WriteIndented = true,
            Converters = { new JsonStringEnumConverter() }
        };

        private readonly object _gate = new();
        private List<WallpaperItem> _items = new();
        private List<BlacklistEntry> _blacklist = new();

        public event Action Changed;

        public void Load()
        {
            lock (_gate)
            {
                _items = ReadList<WallpaperItem>(AppPaths.LibraryFile);
                _blacklist = ReadList<BlacklistEntry>(AppPaths.BlacklistFile);
            }
        }

        private static List<T> ReadList<T>(string path)
        {
            try
            {
                if (File.Exists(path))
                {
                    var json = File.ReadAllText(path);
                    return JsonSerializer.Deserialize<List<T>>(json, Options) ?? new List<T>();
                }
            }
            catch (Exception ex)
            {
                Logger.Error($"Failed to read {Path.GetFileName(path)}.", ex);
            }
            return new List<T>();
        }

        private void SaveItems() => AtomicWrite.WriteAllText(AppPaths.LibraryFile, JsonSerializer.Serialize(_items, Options));
        private void SaveBlacklist() => AtomicWrite.WriteAllText(AppPaths.BlacklistFile, JsonSerializer.Serialize(_blacklist, Options));

        // ---- Queries -------------------------------------------------------

        /// <summary>History newest first.</summary>
        public List<WallpaperItem> GetHistory()
        {
            lock (_gate) return _items.OrderByDescending(i => i.DownloadedAt).ToList();
        }

        public WallpaperItem GetById(string id)
        {
            lock (_gate) return _items.FirstOrDefault(i => i.Id == id);
        }

        public bool IsBlacklisted(string id, string sha256, string url)
        {
            lock (_gate)
            {
                return _blacklist.Any(b =>
                    (!string.IsNullOrEmpty(id) && b.Id == id) ||
                    (!string.IsNullOrEmpty(sha256) && b.Sha256 == sha256) ||
                    (!string.IsNullOrEmpty(url) && b.SourceUrl == url));
            }
        }

        public bool HasItem(string id)
        {
            lock (_gate) return _items.Any(i => i.Id == id);
        }

        public bool HasHash(string sha256)
        {
            lock (_gate) return !string.IsNullOrEmpty(sha256) && _items.Any(i => i.Sha256 == sha256);
        }

        public HashSet<string> UsedIds()
        {
            lock (_gate) return _items.Select(i => i.Id).ToHashSet();
        }

        public List<BlacklistEntry> GetBlacklist()
        {
            lock (_gate) return _blacklist.OrderByDescending(b => b.AddedAt).ToList();
        }

        // ---- Mutations -----------------------------------------------------

        public void AddOrUpdate(WallpaperItem item)
        {
            lock (_gate)
            {
                var existing = _items.FirstOrDefault(i => i.Id == item.Id);
                if (existing != null) _items.Remove(existing);
                _items.Add(item);
                SaveItems();
            }
            Changed?.Invoke();
        }

        public void MarkShown(string id, DateTime when)
        {
            lock (_gate)
            {
                var it = _items.FirstOrDefault(i => i.Id == id);
                if (it == null) return;
                it.LastShownAt = when;
                SaveItems();
            }
            Changed?.Invoke();
        }

        public void SetFavorite(string id, bool favorite)
        {
            lock (_gate)
            {
                var it = _items.FirstOrDefault(i => i.Id == id);
                if (it == null) return;
                it.IsFavorite = favorite;
                SaveItems();
            }
            Changed?.Invoke();
        }

        /// <summary>Remove a history item and (optionally) its files.</summary>
        public void Remove(string id, bool deleteFiles)
        {
            lock (_gate)
            {
                var it = _items.FirstOrDefault(i => i.Id == id);
                if (it == null) return;
                _items.Remove(it);
                if (deleteFiles) TryDeleteFiles(it);
                SaveItems();
            }
            Changed?.Invoke();
        }

        /// <summary>Blacklist an item so it never returns, and delete its files.</summary>
        public void Blacklist(WallpaperItem item)
        {
            lock (_gate)
            {
                if (!_blacklist.Any(b => b.Id == item.Id && !string.IsNullOrEmpty(item.Id)))
                {
                    _blacklist.Add(new BlacklistEntry
                    {
                        Id = item.Id,
                        Sha256 = item.Sha256,
                        SourceUrl = item.SourceUrl,
                        Title = item.Title,
                        AddedAt = DateTime.Now
                    });
                    SaveBlacklist();
                }

                var it = _items.FirstOrDefault(i => i.Id == item.Id);
                if (it != null)
                {
                    _items.Remove(it);
                    TryDeleteFiles(it);
                    SaveItems();
                }
            }
            Changed?.Invoke();
        }

        public void RemoveFromBlacklist(string id, string sha256, string url)
        {
            lock (_gate)
            {
                _blacklist.RemoveAll(b =>
                    (b.Id == id && !string.IsNullOrEmpty(id)) ||
                    (b.Sha256 == sha256 && !string.IsNullOrEmpty(sha256)) ||
                    (b.SourceUrl == url && !string.IsNullOrEmpty(url)));
                SaveBlacklist();
            }
            Changed?.Invoke();
        }

        /// <summary>
        /// Keep the newest <paramref name="maxNonFavorites"/> non-favorite wallpapers on disk;
        /// delete older ones (files + row). Favorites and <paramref name="protectId"/> are never removed.
        /// </summary>
        public int Cleanup(int maxNonFavorites, bool keepFavorites, string protectId)
        {
            var removed = 0;
            lock (_gate)
            {
                var nonFav = _items
                    .Where(i => (!i.IsFavorite || !keepFavorites) && i.Id != protectId)
                    .OrderByDescending(i => i.DownloadedAt)
                    .ToList();

                foreach (var it in nonFav.Skip(Math.Max(0, maxNonFavorites)))
                {
                    _items.Remove(it);
                    TryDeleteFiles(it);
                    removed++;
                }

                if (removed > 0) SaveItems();
            }
            if (removed > 0) Changed?.Invoke();
            return removed;
        }

        private static void TryDeleteFiles(WallpaperItem it)
        {
            foreach (var p in new[] { it.FilePath, it.ThumbnailPath })
            {
                try { if (!string.IsNullOrEmpty(p) && File.Exists(p)) File.Delete(p); }
                catch (Exception ex) { Logger.Warn($"Could not delete {p}: {ex.Message}"); }
            }
        }
    }
}

File: kde/rwr/library.py
"""Wallpaper history plus the blacklist, persisted as JSON. Thread-safe."""
from __future__ import annotations

import json
import threading
from dataclasses import asdict, dataclass, field
from pathlib import Path
from typing import List, Optional

from . import logger, paths


@dataclass
class WallpaperItem:
    id: str
    title: str = ""
    subreddit: str = ""
    author: str = ""
    permalink: str = ""
    source_url: str = ""
    file_path: str = ""
    thumb_path: str = ""
    width: int = 0
    height: int = 0
    file_bytes: int = 0
    sha256: str = ""
    downloaded_at: float = 0.0
    last_shown_at: float = 0.0
    favorite: bool = False

    def file_exists(self) -> bool:
        return bool(self.file_path) and Path(self.file_path).is_file()


@dataclass
class BlacklistEntry:
    id: str = ""
    sha256: str = ""
    source_url: str = ""
    title: str = ""
    added_at: float = 0.0


def _read_list(path: Path, cls):
    try:
        if path.exists():
            raw = json.loads(path.read_text(encoding="utf-8"))
            known = {f for f in cls().__dict__} if _has_default_ctor(cls) else None
            out = []
            for row in raw:
                if known is not None:
                    row = {k: v for k, v in row.items() if k in known}
                out.append(cls(**row))
            return out
    except Exception as ex:
        logger.error(f"Failed to read {path.name}: {ex}")
    return []


def _has_default_ctor(cls) -> bool:
    try:
        cls()
        return True
    except TypeError:
        return False


def _write_list(path: Path, items) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp = path.with_suffix(path.suffix + ".tmp")
    tmp.write_text(json.dumps([asdict(i) for i in items], indent=2), encoding="utf-8")
    tmp.replace(path)


class Library:
    def __init__(self):
        self._lock = threading.RLock()
        self._items: List[WallpaperItem] = []
        self._blacklist: List[BlacklistEntry] = []
        self._changed_listeners = []

    def add_changed_listener(self, fn):
        self._changed_listeners.append(fn)

    def _notify(self):
        for fn in list(self._changed_listeners):
            try:
                fn()
            except Exception:
                pass

    def load(self):
        with self._lock:
            self._items = _read_list(paths.LIBRARY_FILE, WallpaperItem)
            self._blacklist = _read_list(paths.BLACKLIST_FILE, BlacklistEntry)

    # ---- queries ------------------------------------------------------
    def history(self) -> List[WallpaperItem]:
        with self._lock:
            return sorted(self._items, key=lambda i: i.downloaded_at, reverse=True)

    def get(self, item_id: str) -> Optional[WallpaperItem]:
        with self._lock:
            return next((i for i in self._items if i.id == item_id), None)

    def used_ids(self):
        with self._lock:
            return {i.id for i in self._items}

    def has_hash(self, sha: str) -> bool:
        with self._lock:
            return bool(sha) and any(i.sha256 == sha for i in self._items)

    def is_blacklisted(self, item_id="", sha="", url="") -> bool:
        with self._lock:
            return any(
                (item_id and b.id == item_id)
                or (sha and b.sha256 == sha)
                or (url and b.source_url == url)
                for b in self._blacklist
            )

    def blacklist_entries(self) -> List[BlacklistEntry]:
        with self._lock:
            return sorted(self._blacklist, key=lambda b: b.added_at, reverse=True)

    # ---- mutations ----------------------------------------------------
    def add_or_update(self, item: WallpaperItem):
        with self._lock:
            self._items = [i for i in self._items if i.id != item.id]
            self._items.append(item)
            _write_list(paths.LIBRARY_FILE, self._items)
        self._notify()

    def mark_shown(self, item_id: str, when: float):
        with self._lock:
            it = self.get(item_id)
            if not it:
                return
            it.last_shown_at = when
            _write_list(paths.LIBRARY_FILE, self._items)
        self._notify()

    def set_favorite(self, item_id: str, fav: bool):
        with self._lock:
            it = self.get(item_id)
            if not it:
                return
            it.favorite = fav
            _write_list(paths.LIBRARY_FILE, self._items)
        self._notify()

    def remove(self, item_id: str, delete_files: bool):
        with self._lock:
            it = self.get(item_id)
            if not it:
                return
            self._items.remove(it)
            if delete_files:
                _delete_files(it)
            _write_list(paths.LIBRARY_FILE, self._items)
        self._notify()

    def blacklist(self, item: WallpaperItem):
        import time
        with self._lock:
            if item.id and not any(b.id == item.id for b in self._blacklist):
                self._blacklist.append(BlacklistEntry(
                    id=item.id, sha256=item.sha256, source_url=item.source_url,
                    title=item.title, added_at=time.time()))
                _write_list(paths.BLACKLIST_FILE, self._blacklist)
            existing = self.get(item.id)
            if existing:
                self._items.remove(existing)
                _delete_files(existing)
                _write_list(paths.LIBRARY_FILE, self._items)
        self._notify()

    def remove_from_blacklist(self, item_id="", sha="", url=""):
        with self._lock:
            self._blacklist = [
                b for b in self._blacklist
                if not ((item_id and b.id == item_id)
                        or (sha and b.sha256 == sha)
                        or (url and b.source_url == url))
            ]
            _write_list(paths.BLACKLIST_FILE, self._blacklist)
        self._notify()

    def cleanup(self, max_non_favorites: int, keep_favorites: bool, protect_id: str) -> int:
        removed = 0
        with self._lock:
            non_fav = sorted(
                [i for i in self._items
                 if (not i.favorite or not keep_favorites) and i.id != protect_id],
                key=lambda i: i.downloaded_at, reverse=True)
            for it in non_fav[max(0, max_non_favorites):]:
                self._items.remove(it)
                _delete_files(it)
                removed += 1
            if removed:
                _write_list(paths.LIBRARY_FILE, self._items)
        if removed:
            self._notify()
        return removed


def _delete_files(it: WallpaperItem):
    for p in (it.file_path, it.thumb_path):
        try:
            if p and Path(p).is_file():
                Path(p).unlink()
        except Exception as ex:
            logger.warn(f"Could not delete {p}: {ex}")
